plugin/src/compilation/decompiler.ml

```diff
@@ -0,0 +1,89 @@
+
+open Names
+open Constr
+open Environ
+open Envutils
+open Tactics
+open Pp
+open Contextutils
+
+(* Abstraction of Coq tactics supported by this decompiler. *)
+type tact =
+  | Intro of Id.t
+  | Intros of Id.t list
+  | Apply of env * types
+        
+(* Return the string representation of a single tactic.
+   TODO: Indentation/bulleting
+         Don    't expose this in .mli *)
+let show_tact sigma tac : Pp.t =
+  (match tac with
+   | Intro n -> str ("intro " ^ Id.to_string n)
+   | Intros ns ->
+      let names = String.concat " " (List.map Id.to_string ns) in
+      str ("intros " ^ names)
+   | Apply (env, trm) ->
+      let body_s = Printer.pr_constr_env env sigma trm in
+      str "apply " ++ body_s)
+  ++ str ".\n" (* maybe ";" in future *)
+
+(* Converts "intro n. intro m. ..." into "intros n m ..." *)
+let rec collapse_intros (tacs : tact list) : tact list =
+  (*match tacs with
+  | [] -> tacs  
+  | (Intro n) :: (Intro m) :: xs ->
+     collapse_intros (Intros [n; m] :: xs)
+  | (Intros ns) :: (Intro n) :: xs ->
+     collapse_intros (Intros (List.append ns [n]) :: xs)
+  | x :: xs -> x :: collapse_intros xs *)
+  List.fold_right (fun tac acc ->
+      match tac with
+      | Intro n ->
+         (match acc with
+          | Intro m :: xs -> Intros [n; m] :: xs
+          | Intros ns :: xs -> Intros (n :: ns) :: xs
+          | _ -> Intro n :: acc)
+      | t -> t :: acc) tacs []
+
+(* Finds all rel names pushed onto an environment. *)
+let get_pushed_names env =
+  let names =
+    List.map (fun x ->
+        match x with
+        | CRD.LocalAssum (n, _) -> n
+        | CRD.LocalDef (n, _, _) -> n
+      ) (lookup_all_rels env) in
+  Id.Set.of_list
+    (List.map (fun x ->
+         match x with (* shouldn't be anonymous *)
+         | Name n -> n) names)
```

Comment:
To avoid a warning here, you may want to match the other case and raise an exception.